File: crunchy-core/src/main/java/io/crunchy/core/export/MarkdownExporter.java
package io.crunchy.core.export;

import io.crunchy.core.model.CompressedContext;

import java.util.List;

/**
 * Renders the CIR as compact Markdown, ready to paste into an agent prompt.
 *
 * <p>{@code ignoredContent} is deliberately omitted: it is a diagnostic for
 * humans auditing what was removed, and an agent should not spend context
 * reading "3 greetings removed". It remains available in the JSON CIR.
 */
public final class MarkdownExporter {

    public String export(CompressedContext c) {
        StringBuilder md = new StringBuilder();
        md.append("# ").append(c.getIssue()).append("\n\n");
        if (!c.getSummary().isBlank()) {
            md.append(c.getSummary()).append("\n\n");
        }
        if (!c.getBusinessGoal().isBlank()) {
            md.append("**Business goal:** ").append(c.getBusinessGoal()).append("\n\n");
        }
        section(md, "Architecture", c.getArchitecture());
        section(md, "Decisions", c.getDecisions());
        section(md, "Constraints", c.getConstraints());
        section(md, "Acceptance Criteria", c.getAcceptanceCriteria());
        section(md, "Risks", c.getRisks());
        section(md, "TODOs", c.getTodos());
        section(md, "Dependencies", c.getDependencies());
        section(md, "Related Issues", c.getRelatedIssues());
        md.append("---\n")
          .append("confidence: ").append(c.getConfidence())
          .append(" | compression: ").append(Math.round(c.getCompressionRatio() * 100)).append("%\n");
        return md.toString();
    }

    private void section(StringBuilder md, String title, List<String> items) {
        if (items.isEmpty()) {
            return;
        }
        md.append("## ").append(title).append("\n");
        for (String item : items) {
            md.append("- ").append(item).append("\n");
        }
        md.append("\n");
    }
}

File: web/src/core/export/MarkdownExporter.ts
import { CompressedContext } from '../model/CompressedContext.js';

/**
 * Renders the CIR as compact Markdown, ready to paste into an agent prompt.
 *
 * `ignoredContent` is deliberately omitted: it is a diagnostic for humans
 * auditing what was removed, and an agent should not spend context reading
 * "3 greetings removed". It remains available in the JSON CIR.
 */
export function toMarkdown(c: CompressedContext): string {
  const md: string[] = [];
  md.push(`# ${c.issue}\n`);
  if (c.summary.trim().length > 0) {
    md.push(`${c.summary}\n`);
  }
  if (c.businessGoal.trim().length > 0) {
    md.push(`**Business goal:** ${c.businessGoal}\n`);
  }
  section(md, 'Architecture', c.architecture);
  section(md, 'Decisions', c.decisions);
  section(md, 'Constraints', c.constraints);
  section(md, 'Acceptance Criteria', c.acceptanceCriteria);
  section(md, 'Risks', c.risks);
  section(md, 'TODOs', c.todos);
  section(md, 'Dependencies', c.dependencies);
  section(md, 'Related Issues', c.relatedIssues);
  md.push('---');
  md.push(
    `confidence: ${c.confidence} | compression: ${Math.round(c.compressionRatio * 100)}%`,
  );
  return md.join('\n') + '\n';
}

function section(md: string[], title: string, items: string[]): void {
  if (items.length === 0) {
    return;
  }
  md.push(`## ${title}`);
  for (const item of items) {
    md.push(`- ${item}`);
  }
  md.push('');
}
